src/preprocessing/numerical.rs

```diff
@@ -0,0 +1,406 @@
+/// neighbors based models.
+#[derive(Clone, Debug, Default, PartialEq)]
+struct StandardScaler<T: RealNumber> {
+    means: Vec<T>,
+    stds: Vec<T>,
+    parameters: StandardScalerParameters,
+}
+impl<T: RealNumber> StandardScaler<T> {
+    /// When the mean should be adjusted, the column mean
+    /// should be kept. Otherwise, replace it by zero.
+    fn adjust_column_mean(&self, mean: T) -> T {
+        if self.parameters.with_mean {
+            mean
+        } else {
+            T::zero()
+        }
+    }
+    /// When the standard-deviation should be adjusted, the column
+    /// standard-deviation should be kept. Otherwise, replace it by zero.
+    fn adjust_column_std(&self, std: T) -> T {
+        if self.parameters.with_std {
+            ensure_std_valid(std)
+        } else {
+            T::one()
+        }
+    }
+}
+
+/// Make sure the standard deviation is valid. If it is
+/// negative or zero, it should replaced by the smallest
+/// positive value the type can have. That way we can savely
+/// divide the columns with the resulting scalar.
+fn ensure_std_valid<T: RealNumber>(value: T) -> T {
+    value.max(T::min_positive_value())
+}
+
+/// During `fit` the `StandardScaler` computes the column means and standard deviation.
+impl<T: RealNumber, M: Matrix<T>> UnsupervisedEstimator<M, StandardScalerParameters>
+    for StandardScaler<T>
+{
+    fn fit(x: &M, parameters: StandardScalerParameters) -> Result<Self, Failed> {
+        Ok(Self {
+            means: x.column_mean(),
+            stds: x.std(0),
+            parameters,
+        })
+    }
+}
+
+/// During `transform` the `StandardScaler` applies the summary statistics
+/// computed during `fit` to set the mean of each column to zero and the
+/// standard deviation to one.
+impl<T: RealNumber, M: Matrix<T>> Transformer<M> for StandardScaler<T> {
+    fn transform(&self, x: &M) -> Result<M, Failed> {
+        let (_, n_cols) = x.shape();
+        if n_cols != self.means.len() {
+            return Err(Failed::because(
+                FailedError::TransformFailed,
+                &format!(
+                    "Expected {} columns, but got {} columns instead.",
+                    self.means.len(),
+                    n_cols,
+                ),
+            ));
+        }
+
+        Ok(build_matrix_from_columns(
+            self.means
+                .iter()
+                .zip(self.stds.iter())
+                .enumerate()
+                .map(|(column_index, (column_mean, column_std))| {
+                    take_column_from_matrix(x, column_index)
+                        .sub_scalar(self.adjust_column_mean(*column_mean))
+                        .div_scalar(self.adjust_column_std(*column_std))
+                })
+                .collect(),
+        )
+        .unwrap())
+    }
+}
+
+/// Extract a single column from a matrix and return it as a matrix.
+fn take_column_from_matrix<T, M>(matrix: &M, column_index: usize) -> M
```

Comment:
These two function look really useful, maybe you should consider adding them to `linalg::BaseMatrix` implementation as they are more general than the scaler implementation.
It would be nice to have something like `DenseMatrix::take_column(...)` (I thought we already have something like it?)

Reply:
there is already a `take` implementation for matrices.

Reply:
yes exactly, I also use the `Matrix::take` interface here, but I want to specialize it for taking columns (then you need to paramterize it via axis, e.g. `Matrix::take(&[column_index], 1)`). This is helpful because then I don't need to repeat this in my code and can test it in a central place (I always like to mess up the axis indices :laughing: ). So if it is ok, I would still like to keep `build_matrix_from_columns` (or put it in the `Matrix`-implementation).

Apart from this I would also be happy to put any of the two methods (`build_matrix_from_columns` | `take_column`) into the `Matrix` imlementation. Should I do it? For which of the two methods?

Reply:
If they have generic functionalities and they can be reused, not strictly tied to the implementation of the module, they can be moved to modules with a more generic functionalities so that other modules can make use of them. I think you can move `take_columns_from_matrix` into `linalg::BaseMatrix` implementation so it can be used like `BaseMatrix::take_columns_from_matrix`; same for `build_matrix_from_columns`.

Reply:
As we discussed below, I think `take_column_from_matrix` , or rather `Matrix:take_column` (what you proposed earlier) I would put it into a more central place as it can be used in other places as well.